### ISSUE TITLE
Tracking issue: Image Lightbox

### ISSUE THREAD

**Issue by artemiomorales** [WordPress/gutenberg#51132]
This tracking issue is for identifying bugs and features related to the experimental lightbox, introduced in #50373.

#### MVP
- [x] https://github.com/WordPress/gutenberg/issues/50959
- [x] https://github.com/WordPress/gutenberg/issues/50973
- [x] https://github.com/WordPress/gutenberg/issues/51068
- [x] https://github.com/WordPress/gutenberg/issues/51125
- [x] https://github.com/WordPress/gutenberg/issues/51126
- [x] https://github.com/WordPress/gutenberg/issues/51127
- [x] https://github.com/WordPress/gutenberg/issues/51170
- [x] https://github.com/WordPress/gutenberg/issues/51171
- [x] https://github.com/WordPress/gutenberg/issues/51276
- [x] https://github.com/WordPress/gutenberg/pull/51305#pullrequestreview-1469318233
  - Adding a link to image will disable lightbox
  - Ensure lightbox still works when adding an image via external URL

#### v0.1
**Goal: Provide a robust, stable implementation supporting all single image features with UI exposed publicly to users for WordPress 6.4**
**Target Deadline: 17 October 2023**
- [x] https://github.com/WordPress/gutenberg/issues/51555
- [x] https://github.com/WordPress/gutenberg/issues/51055
- [x] https://github.com/WordPress/gutenberg/pull/51748
- [x] https://github.com/WordPress/gutenberg/pull/51692
- [x] https://github.com/WordPress/gutenberg/issues/51556
- [x] https://github.com/WordPress/gutenberg/pull/52067
- [x] https://github.com/WordPress/gutenberg/pull/52187
- [x] https://github.com/WordPress/gutenberg/pull/52140
- [x] https://github.com/WordPress/gutenberg/issues/52183
- [x] #52172
- [x] https://github.com/WordPress/gutenberg/pull/52290
- [x] https://github.com/WordPress/gutenberg/pull/52833
- [x] https://github.com/WordPress/gutenberg/pull/52765
- - [x] [Use transform or the new translate](https://github.com/WordPress/gutenberg/pull/51357#discussion_r1233549304) prop for more performant animations
- [x] https://github.com/WordPress/gutenberg/issues/53403
- - [x] https://github.com/WordPress/gutenberg/pull/53851
- - [x] https://github.com/WordPress/gutenberg/pull/54071
- [x] https://github.com/WordPress/gutenberg/pull/54878
- [x] https://github.com/WordPress/gutenberg/pull/54639
- [x] https://github.com/WordPress/gutenberg/pull/54670
- [x] https://github.com/WordPress/gutenberg/pull/53026
- [x] https://github.com/WordPress/gutenberg/issues/54930
- [x] https://github.com/WordPress/gutenberg/issues/54921
- [x] https://github.com/WordPress/gutenberg/issues/54869
- [x] https://github.com/WordPress/gutenberg/pull/55000
- [x] https://github.com/WordPress/gutenberg/issues/54971
- [x] https://github.com/WordPress/gutenberg/pull/55077
- [x] https://github.com/WordPress/gutenberg/pull/52442
- [x] https://github.com/WordPress/gutenberg/issues/55119#issuecomment-1751101370
- [x] https://github.com/WordPress/gutenberg/pull/55141
- [x] https://github.com/WordPress/gutenberg/issues/54940#issuecomment-1743339787
- [x] https://github.com/WordPress/gutenberg/issues/55097
- [x] https://github.com/WordPress/gutenberg/issues/55214#issuecomment-1756819319
- [x] https://github.com/WordPress/gutenberg/issues/54682
- [x] https://github.com/WordPress/gutenberg/issues/55414
- [x] https://github.com/WordPress/gutenberg/issues/55462
- [x] https://github.com/WordPress/gutenberg/pull/55724

### WordPress 6.5 Maintenance
- [x] https://github.com/WordPress/gutenberg/pull/57608

#### v0.2

- [ ] https://github.com/WordPress/gutenberg/issues/57849
- [ ] https://github.com/WordPress/gutenberg/issues/55617
- [ ] https://github.com/WordPress/gutenberg/issues/55088
- [x] https://github.com/WordPress/gutenberg/issues/55123#event-10574514151
- [ ] https://github.com/WordPress/gutenberg/issues/54858
- [ ] Discuss changing lightbox button placement for `rtl` languages
- [ ] Add tests for new UI

### Backlog

- [ ] [Explore using lightbox as a replacement for attachment pages](https://core.trac.wordpress.org/ticket/57913#comment:27)
- [ ] Perform discovery around adding support for additional blocks
- [ ] https://github.com/WordPress/gutenberg/issues/54635
- - Featured image?
- - Cover?
- [ ] Represent a preview of the currently selected lightbox animation effect in the sidebar
- [ ] Add support for variable mobile browser window height due to address bar being visible or not
- [ ] Clean up directives using Tag Processor bookmarks and HTML Processor
- [x] https://github.com/WordPress/gutenberg/pull/43059#issuecomment-1562240523
- [ ] `Resize()` callback for lightbox button does not always fire, causing horizontal overflow
- [ ] Add callback on window `resize()` to recenter lightbox when it is active
- [ ] Explore extending lightbox to include support for additional content types
- [ ] Explore extending lightbox to include support for galleries
- [ ] [Add zoom on hover](https://github.com/WordPress/gutenberg/issues/51055#issuecomment-1574849580)
- [ ] Disable lightbox when it is part of a gallery
- [ ] Explore using CSS background instead of JavaScript to improve animation performance
  - Potential idea around Accelerated Mobile Pages and how it used `width` and `height` properties to perform layout adjustments before loading media
  - [Example of loading an image as a CSS background](https://xkq.io/blog/cowboy-nitrogen-install)

**Comment by gziolo** [comment]
Sharing more issues, I reported today:

- https://github.com/WordPress/gutenberg/issues/51125
- https://github.com/WordPress/gutenberg/issues/51126
- https://github.com/WordPress/gutenberg/issues/51127

**Comment by ajlende** [comment]
Added one more that I found

- https://github.com/WordPress/gutenberg/issues/51276

**Comment by michalczaplinski** [comment]
I've opened the PR to: 
- https://github.com/WordPress/gutenberg/pull/52140

**Comment by cbravobernal** [comment]
I've updated the issue description with these new PRs
 - https://github.com/WordPress/gutenberg/pull/52833
 - https://github.com/WordPress/gutenberg/pull/52765
 - https://github.com/WordPress/gutenberg/pull/52829

**Comment by artemiomorales** [comment]
> I've updated the issue description with these new PRs
> - https://github.com/WordPress/gutenberg/pull/52833
> - https://github.com/WordPress/gutenberg/pull/52765
> - https://github.com/WordPress/gutenberg/pull/52829

@c4rl0sbr4v0 Great, thanks! Heads up that I moved #52765 to [Interactivity Behaviors MVP](https://github.com/WordPress/gutenberg/issues/52292) instead, since this issue is more a catch-all for all tasks related to the lightbox specifically, though am happy to discuss and iterate if there's a better approach.

**Comment by artemiomorales** [comment]
## Changes in Scope for v0.1

We've added a few issues to the v0.1 release, with a new target deadline of **25 August 2023**. You can find more details in the dedicated [Interactivity API Behaviors MVP tracking issue](https://github.com/WordPress/gutenberg/issues/52292).

## New Scope for v0.2

Planning for WordPress 6.4 has also kicked off, and I've begun outlining plans for a v0.2 inclusion of the lightbox.

I've set the target goal to be the following: **Provide a robust, stable implementation supporting all single image features in WordPress 6.4.**

We'll need to perform discovery around some proposed additions to the scope, including the tasks to be done for adding lightbox support for the cover block, featured image, and potentially other blocks.

As of now, that deadline has been set for **22 September 2023**, the last Gutenberg release before the 6.4 beta.

That timeline and feature set as proposed by design honestly seems a bit tight at the moment. That being said, I'm happy to hear thoughts and further discuss the items for inclusion in 6.4 and this update overall. Thanks!

**Comment by artemiomorales** [comment]
### Biweekly Update

The lightbox was originally envisioned as an MVP of an idea called [behaviors](https://github.com/WordPress/gutenberg/issues/50029), but since it appears increasingly that the lightbox is a particular setting of the _image block_, we will no longer be shipping behaviors as a concept for WordPress 6.4. This means that the [tracking issue for the lightbox as a behaviors MVP](https://github.com/WordPress/gutenberg/issues/50029) has now been closed.

This also means that we need to revise the lightbox UI, which is currently being discussed in [Image Block: Revise Lightbox UI to Remove Reference to Behaviors](https://github.com/WordPress/gutenberg/issues/53403).

We also finally merged [aspect ratio support](https://github.com/WordPress/gutenberg/pull/52765), which means that the lightbox now supports all single image block features.

With that in mind, implementing the new lightbox UI takes precedence over adding support for additional block types and and as a potential replacement for the attachment page, so I've revised the v0.1 and v0.2 goals and scope to reflect that prioritization. 

In other words, the v0.1 will be the primary focus until it's shipped, with the v0.2 tasks being an aspirational goal for WordPress 6.4. I'll continue updating the v0.2 list of tasks as we complete the UI and the scope for those tasks becomes better defined.

#### Merged
- https://github.com/WordPress/gutenberg/pull/52765

#### Closed
- https://github.com/WordPress/gutenberg/issues/52292

**Comment by artemiomorales** [comment]
### Biweekly Update

These past couple of weeks we've spent a lot of time discussing what the UI for the lightbox should look like for the release of the WP 6.4 beta, slated for 26 September 2023, captured in the [following issue](https://github.com/WordPress/gutenberg/issues/53403). All feedback considered, the current design we're working on is captured in [this comment](https://github.com/WordPress/gutenberg/issues/53403#issuecomment-1697372363).

We have two open PRs to remove the old Behaviors UI and introduce new settings panels for the image in the Global Styles and block settings, which have been added to the task list for v0.1:
- https://github.com/WordPress/gutenberg/pull/53851
- https://github.com/WordPress/gutenberg/pull/54071

We're aiming to have these outstanding tasks completed by 13 September 2023, the last Gutenberg release so we can comfortably have a couple of weeks of testing before the lightbox is, if all goes well, merged to WordPress core.

**Comment by annezazu** [comment]
@artemiomorales where do we stand on this? I don't see a recent update in the last two weeks and I'd like to feature this in a next call for testing for the FSE Outreach Program :)

**Comment by artemiomorales** [comment]
@annezazu Thanks for reaching out! As far as I can see, the lightbox will be included in Gutenberg 16.7 and WP 6.4, so yes  having more eyes on it would be great 😄 Here follows a more detailed update:

### Biweekly Update

We've removed the old UI for the lightbox and added its new UI as an exclusive setting for the image block. This means that the lightbox, barring any unforeseen circumstances, will be shipped in WordPress 6.4.

While implementing the UI changes, we noticed some inconsistency in how `theme.json` settings are accessed and parsed in the codebase, so while the new UI will work without issue in a majority of cases, certain `theme.json` values will cause unexpected behavior under certain circumstances.

As beta testing commences, we'll continue to refine this and fix any other bugs that arise throughout the beta period over the next few weeks, as well as add tests. These new action items have been added to the v0.1 scope.

Merged
- https://github.com/WordPress/gutenberg/pull/53851
- https://github.com/WordPress/gutenberg/pull/54071

Closed
- https://github.com/WordPress/gutenberg/issues/53403

**Comment by bph** [comment]
@artemiomorales It all works fine. h owever, maybe you can elaborate on this with one or two examples (or point to the educating issue)

> certain theme.json values will cause unexpected behavior under certain circumstances.

**Comment by artemiomorales** [comment]
@bph Thank you for taking a look at this! We've captured the `theme.json` inconsistencies in the following issues. The v0.1 task list has also been updated with these links.

- https://github.com/WordPress/gutenberg/issues/54858
- https://github.com/WordPress/gutenberg/issues/54635
- https://github.com/WordPress/gutenberg/issues/54638

**Comment by annezazu** [comment]
Marking this as blessed for [major release per this slack discussion](https://wordpress.slack.com/archives/C055Y7FKS7N/p1695052825968849). @artemiomorales can you confirm that would be desirable? It seems like we'll likely get everything in in time for beta :) We can easily remove the status and we do try to keep those marked as blessed small in number.

**Comment by artemiomorales** [comment]
@annezazu Yes that sounds good! We'll likely merge a couple of more enhancements before the beta period is over, so this sounds right.

**Comment by artemiomorales** [comment]
### Biweekly Update

These past two weeks, we've been reporting and fixing bugs to prepare for WordPress 6.4 — a huge thanks to everyone who has been helping with testing 🙏

Worth noting: We've clarified the scope regarding which `theme.json` issues we'll address and fixed a layout shift occurring with the zoom.

Specifically with the `theme.json`, for now the plan is to add top-level setting support for the lightbox but **not** add support for the shorthand syntax.

Given the newly reported bugs, I've updated the target date for v0.1 to be **October 17th**, the projected date for the WordPress 6.4 Release Candidate 1. I've added the new bugs to the scope.

**Merged**

- https://github.com/WordPress/gutenberg/pull/54878
- https://github.com/WordPress/gutenberg/pull/54639
- https://github.com/WordPress/gutenberg/pull/54670
- https://github.com/WordPress/gutenberg/pull/53026

**Bugs added to the v0.1 scope**
- https://github.com/WordPress/gutenberg/issues/54921
- https://github.com/WordPress/gutenberg/issues/54869
- https://github.com/WordPress/gutenberg/issues/54916
- https://github.com/WordPress/gutenberg/issues/54682

The plan is to continue addressing bugs, do more testing, and add automated tests.

**Comment by artemiomorales** [comment]
### Biweekly Update

These past two weeks we've been wrapping up outstanding bugs to get the lightbox included in 6.4. Having more eyes on this continues to be great as folks chip in with identifying or fixing issues, from accessibility to backwards compatibility and more.

**Closed / Merged**

- https://github.com/WordPress/gutenberg/issues/54971
  - https://github.com/WordPress/gutenberg/pull/55010
- https://github.com/WordPress/gutenberg/issues/55124
  - https://github.com/WordPress/gutenberg/pull/55077
- https://github.com/WordPress/gutenberg/issues/55119
  - https://github.com/WordPress/gutenberg/pull/55120
- https://github.com/WordPress/gutenberg/issues/54940#issuecomment-1743339787

**Remaining bugs for v0.1 scope**
- https://github.com/WordPress/gutenberg/issues/55097 (needed for 6.4)
  - https://github.com/WordPress/gutenberg/issues/54682 (may resolve above)
- https://github.com/WordPress/gutenberg/issues/54858
- https://github.com/WordPress/gutenberg/issues/55123#event-10574514151
- https://github.com/WordPress/gutenberg/issues/55088 
- Add tests

**Comment by JosKlever** [comment]
It looks like I've found a bug, but I'm not sure yet how it works. I've read a post using the new lightbox, and instead of showing the lightbox icon in the top right corner of an images on hover, it showed on 100% width/height even without the hover covering the all the images in the post with a black overlay.

This happened on both my (Windows) desktop as laptop with Chrome, Chrome incognito and Edge, but not with Firefox. All latest versions. There were no errors in the browser console and no extensions that could have blocked something,

In the end I've tested it without some local security and disabled Malwarbytes Anti-Malware. After that if worked as it should. So Malwarebytes seems to block a script or something that prevents the lightbox to work correctly. 

More research is probably necessary, but I'll have to do that later. If this is not the right place to report it, please let me know where it belongs.

**Comment by artemiomorales** [comment]
> It looks like I've found a bug, but I'm not sure yet how it works. I've read a post using the new lightbox, and instead of showing the lightbox icon in the top right corner of an images on hover, it showed on 100% width/height even without the hover covering the all the images in the post with a black overlay.

@JosKlever Thanks for your comment, and pardon the delay!

Would you be able to [file a bug report](https://github.com/WordPress/gutenberg/issues/new?assignees=&labels=&projects=&template=Bug_report.yml)? That will give the issue more visibility and allow folks to try to reproduce it, discuss, and suggest solutions 😄

**Comment by JosKlever** [comment]
@artemiomorales The issue appears to be solved now. I guess there was an update in the last 10 days that fixed it.

**Comment by artemiomorales** [comment]
### Biweekly Update

I'm a bit late on this update, but it looks like the lightbox has made it to WordPress 6.4.

Once 6.4 is released, we hope to receive more feedback on the implementation and continue iterating our approach.

This seems to be a good milestone, so I think we can declare the v0.1 of the lightbox done; I've moved remaining tasks to v0.2 and will continue developing the scope of work for 6.5 there.

For the moment, these are the next issues we'll be tackling:

- https://github.com/WordPress/gutenberg/issues/55617
- https://github.com/WordPress/gutenberg/issues/54916
- https://github.com/WordPress/gutenberg/issues/55088
- https://github.com/WordPress/gutenberg/issues/54858
- https://github.com/WordPress/gutenberg/issues/55123
- https://github.com/WordPress/gutenberg/issues/56310

Thanks everyone for your help in getting the lightbox to this stage of development! 🙌 🥳

**Comment by artemiomorales** [comment]
### Update

Hey everyone, here I'm shooting an update to let everyone know that I'll be away until the week of December 4th and that biweekly updates may pause until I'm back.

Please reach out to @michalczaplinski on any issues or for any questions and discussions pertaining to the lightbox in the meantime. Thank you!

**Comment by NicoHood** [comment]
Will the lightbox be also added to the text+media block? I did not find it in this roadmap.

**Comment by michalczaplinski** [comment]
There are no short-term plans to do that as far as I'm aware, but I'd very happily review any PRs that add it! It might be a bit tricky because the Media&Text block can have different alignment configurations.

**Comment by artemiomorales** [comment]
### Update

Hello all! Since the last update, my focus has mostly shifted to another project for 6.5, and someone else may be overseeing the lightbox come 6.6.

For the moment, I'm aiming to deliver just an improved UX for the lightbox in 6.5 via the following PR, as well as deal with any critical bugs if they arise:
- https://github.com/WordPress/gutenberg/pull/57608

Please feel free to continue reporting bugs and suggesting improvements; if you mention this issue, I can add those items to the backlog here for reference 🙏

**Comment by mtias** [comment]
Closing this tracking issue since it's not active anymore.
